snuba/query/conditions.py

```diff
@@ -0,0 +1,157 @@
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from enum import Enum
+from typing import Callable, Iterator, Sequence
+
+from snuba.query.collections import NodeContainer
+from snuba.query.expressions import Expression, ExpressionContainer
+from snuba.query.nodes import FormattableNode
+
+
+class Operator(Enum):
+    GT = ">"
+    LT = "<"
+    GTE = ">="
+    LTE = "<="
+    EQ = "="
+    NEQ = "!="
+    IN = "IN"
+    NOT_IN = "NOT IN"
+    IS = "IS"
+    LIKE = "LIKE"
+    NOT_LIKE = "NOT LIKE"
+
+
+class Condition(FormattableNode, ABC):
+    """
+    Represents a condition node in the query. This can be a simple infix
+    notation query or a complex query built by a nested boolean condition.
+
+    """
+    @abstractmethod
+    def get_expressions(self) -> ExpressionContainer:
```

Comment:
This is the trickiest part of this tree.
Here we have a subtree that represent conditions for several levels in the tree. Then at a given level we have Condition leaves (BasicCondition) which on their own are the root of expression trees. So it is not practical (nor useful) to map across both sections of the tree. The only common type is Node. What is more useful is being able to iterate over Expressions across the tree no matter how nested they are, and this is what the ExpressionCOntainer does.